### PR TITLE
Change COMPAT locale references to CLDR

### LINKED ISSUE #109003: JDK 23 changes how locales work in a way that breaks elasticsearch
There's various locale-based test failures when ES tests are run on JDK 23:

https://github.com/elastic/elasticsearch/issues/109004
https://github.com/elastic/elasticsearch/issues/108625
https://github.com/elastic/elasticsearch/issues/108855

Tests are failing because the locale options are not being applied correctly.

However, when the following is run in a completely standalone class on jdk23:

`DateTimeFormatter.ofPattern("eeee", Locale.GERMAN).format(ZonedDateTime.now())`

it does correctly use the locale, and returns the German day name. When the same code is run inside an ES test, it _fails_ to apply the locale and returns the English day name.

### PR DESCRIPTION
**Pull request by thecoop**
The COMPAT locale data provider has been removed in JDK23 by https://github.com/openjdk/jdk/pull/17991. This PR swaps out the old COMPAT locale provider for the default CLDR provider.

As part of this, the differences between CLDR and COMPAT have been encapsulated by the `CompatCalendarNameProvider` class, generated manually (for now) between two diffs of the contents of each locale provider, as given in the javadoc. This is applied by the SPI locale provider, before CLDR is queried.

Although the JDK may have changed the locale behaviour, we still need to maintain the existing behaviour for our customers across minor versions. We can't just change how date formats work in a minor ES release, despite them all actually coming from the JDK - they should work the same regardless of the JDK version that ES runs on.

Deprecation and migration strategies for ESv9+ are still to be determined.

This fixes #109003

### CODE REVIEW

**Comment by thecoop** [comment]
This actually has quite a significant difference - the ROOT locale differs in behaviour between COMPAT and CLDR; one such difference is that it doesn't have long day/month names, only short names, which given we use ROOT for everything has a significant impact on our date formatters

**Review by mark-vieira — APPROVED** [review]

File: distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/SystemJvmOptions.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ static List<String> systemJvmOptions(Settings nodeSettings, final Map<String, St
                  * Due to internationalization enhancements in JDK 9 Elasticsearch need to set the provider to COMPAT otherwise time/date
```
Comment (mark-vieira) [inline distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/SystemJvmOptions.java:68]: Should we update this comment?

**Comment by mark-vieira** [comment]
Does this change need to be conditional based on the runtime JVM or is it safe to just change it for all supported runtimes?

**Comment by thecoop** [comment]
There's also differences in german between COMPAT and CLDR. `E` day-of-week format produces `Mo` on COMPAT and `Mo.` on CLDR. This changes parsing and outputting of dates using that specifier. How are we going to find all of these differences, short of replicating the entirety of COMPAT in the ES codebase?

**Comment by rjernst** [comment]
> short of replicating the entirety of COMPAT in the ES codebase?

I think that is the only way to maintain the COMPAT behavior until we can properly deprecate the behavior.

**Comment by pgomulka** [comment]
can we copy the entier compat locale providers into a small jar that we include as a dependency and remove at some point in the future?

**Comment by thecoop** [comment]
The PR that removed COMPAT is https://github.com/openjdk/jdk/pull/17991, it covers over 500 files. I'm also not clear where the ROOT locale information is stored - I can't find it in the various removed files.

File: server/src/main/java/org/elasticsearch/common/time/CompatCalendarNameProvider.java [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+ * <p>
+ * The source data can be obtained by iterating through {@link Locale#getAvailableLocales()}
+ * for both {@code -Djava.locale.providers=CLDR} and {@code -Djava.locale.providers=COMPAT},
+ * obtaining {@link java.util.Calendar#getInstance} for each locale,
+ * and calling {@code java.util.Calendar#getDisplayNames} for all valid values of {@code field} and {@code style}.
+ * This data can then be written to disk in an appropriate textual format, and compared with a diff tool.
+ * <p>
+ * Locales with a country extension will just have the same english-base info as {@link Locale#ROOT}, and so can be ignored.
+ * We're only concerned about the top-level language differences here.
+ */
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/time/CompatCalendarNameProvider.java:37]: to confirm, that class was generated based on a data described in the javadoc?

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/common/time/CompatCalendarNameProvider.java:37]: This is not auto-generated; I generated the diffs, then created the overrides manually. I wanted to get something working quickly to validate this does actually fix the test failures, and merge it fairly quickly.

We can discuss how formal we want to make this override as part of longer-term support for this class - possibly creating gradle tasks for generating the locale files, doing the diff, creating the overrides file, etc...

**Review by pgomulka — APPROVED** [review]
approved,
I think this is a good approach for short term fix of jdk23 test
perhaps it is even good for long term, assuming we don't care about other  LocaleProviders that might be affected

File: distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/SystemJvmOptions.java (hunk 1 above)
Comment (thecoop) [reply distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/SystemJvmOptions.java:68]: Yes, we should...

**Comment by thecoop** [comment]
@elasticmachine update branch

**Comment by thecoop** [comment]
@elasticmachine update branch

**Comment by thecoop** [comment]
Following discussions within ES, we can't maintain our own locale database independent of CLDR. Closing in favour of #110222

**Pull request closed without merging**